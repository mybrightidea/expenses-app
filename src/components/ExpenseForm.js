import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: undefined
    };
  }

  onFormSubmit = e => {
    e.preventDefault();
    const { description, createdAt, note, amount } = this.state;
    if (!amount || !description) {
      this.setState(() => ({
        error: "Amount and description must be entered"
      }));
    } else {
      this.setState(() => ({
        error: undefined
      }));

      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        note,
        createdAt: createdAt.valueOf()
      });
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = e => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  render() {
    const {
      description,
      note,
      amount,
      createdAt,
      calendarFocused,
      error
    } = this.state;

    return (
      <form className="form" onSubmit={this.onFormSubmit}>
        <input
          type="text"
          placeholder="Description"
          autoFocus
          className="text-input"
          value={description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={this.onDateChange}
          focused={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          displayFormat={"DD/MM/YYYY"}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          className="textarea"
          value={note}
          onChange={this.onNoteChange}
        ></textarea>
        <div>
          <button className="button">{this.props.updateLabel}</button>
        </div>

        {error && <p className="form__error">{error}</p>}
      </form>
    );
  }
}

export default ExpenseForm;
