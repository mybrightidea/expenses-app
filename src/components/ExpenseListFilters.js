import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";
import "react-dates/initialize";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = focused => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onTextChange = e => {
    e.preventDefault();
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = e => {
    e.preventDefault();
    e.target.value === "date"
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };
  render() {
    const { text, sortBy, startDate, endDate } = this.props.filters;
    const { calendarFocused } = this.state;
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search expenses"
              type="text"
              value={text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={startDate}
              endDate={endDate}
              startDateId="startDate"
              endDateId="endDate"
              onDatesChange={this.onDatesChange}
              focusedInput={calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
              displayFormat={"DD/MM/YYYY"}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({ filters: state.filters });

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
