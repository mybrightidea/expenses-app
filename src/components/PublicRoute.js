import React from "react";
import { connect } from "react-redux";
import { Route, Redirect } from "react-router-dom";
import Header from "../components/Header";

export const PublicRoute = props => {
  const { isAuthenticated, component: Component, ...rest } = props;
  return (
    <Route
      {...rest}
      component={props =>
        !isAuthenticated ? (
          <div>
            <Component {...props} />
          </div>
        ) : (
          <Redirect to="/dashboard" />
        )
      }
    />
  );
};

const mapStateToProps = state => ({ isAuthenticated: !!state.auth.uid });

export default connect(mapStateToProps)(PublicRoute);
