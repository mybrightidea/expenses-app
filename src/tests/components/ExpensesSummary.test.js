import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("Should work fine with one value", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={1} expensesTotal={55} />
  );
  expect(wrapper).toMatchSnapshot();
});
test("Should work fine with no value", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={0} expensesTotal={0} />
  );
  expect(wrapper).toMatchSnapshot();
});
test("Should work fine with many value", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={27} expensesTotal={1273521753} />
  );
  expect(wrapper).toMatchSnapshot();
});
