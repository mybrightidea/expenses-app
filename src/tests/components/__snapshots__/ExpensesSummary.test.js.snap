// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should work fine with many value 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Your 
      <span>
        27
      </span>
       
      expenses
       totalling
       
      <span>
        £12,735,217.53
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should work fine with no value 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Your 
      <span>
        0
      </span>
       
      expenses
       totalling
       
      <span>
        £0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should work fine with one value 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Your 
      <span>
        1
      </span>
       
      expense
       totalling
       
      <span>
        £0.55
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
